"Run in Docker" readme section added (#50)
triggermesh/knative-lambda-runtime: README.md

{"message":"Hello, the current time is Tue Apr 07 13:59:17 GMT 2020"}
```

### Run in Docker 

> For cases in which the use of additional components (tm CLI, Tekton, Knative, k8s) is undesirable, it is possible to build KLR function as a standalone Docker container and run it in any environment. To do this, you should extract Dockerfile from the runtime you are interested in, put it in the directory with your function, update the handler variable, and build the container. Here is Dockerfile definition for all runtimes:

Reviewer: ```suggestion
For cases in which the use of additional components (tm CLI, Tekton, Knative, k8s) is undesirable, it is possible to build a KLR function as a standalone Docker container and run it in any environment. To do this, you should extract the Dockerfile from the runtime you are interested in, put it in the directory with your function, update the handler variable, and build the container. Here are Dockerfile definitions for all runtimes:
```

> - [go](https://github.com/triggermesh/knative-lambda-runtime/blob/9a74ce1ac03d56d233cfc7a46d84f2c5e5f2685a/go-1.x/runtime.yaml#L44-L68)

Reviewer 2: these links will break when we make additional commits right ?

Author: not really, that's why I pointed to the particular commit instead of a head - these links will stay relevant despite any changes in the master. We may need to update these links in case of important, incompatible changes, but for now, it should work just fine.

- [java](https://github.com/triggermesh/knative-lambda-runtime/blob/9a74ce1ac03d56d233cfc7a46d84f2c5e5f2685a/java8/runtime.yaml#L43-L53)
- [node-10](https://github.com/triggermesh/knative-lambda-runtime/blob/9a74ce1ac03d56d233cfc7a46d84f2c5e5f2685a/node-10.x/runtime.yaml#L43-L48)
- [node-4](https://github.com/triggermesh/knative-lambda-runtime/blob/9a74ce1ac03d56d233cfc7a46d84f2c5e5f2685a/node-4.x/runtime.yaml#L43-L48)
- [python-2](https://github.com/triggermesh/knative-lambda-runtime/blob/9a74ce1ac03d56d233cfc7a46d84f2c5e5f2685a/python-2.7/runtime.yaml#L43-L50)
- [python-3](https://github.com/triggermesh/knative-lambda-runtime/blob/9a74ce1ac03d56d233cfc7a46d84f2c5e5f2685a/python-3.7/runtime.yaml#L43-L50)
- [ruby-2](https://github.com/triggermesh/knative-lambda-runtime/blob/9a74ce1ac03d56d233cfc7a46d84f2c5e5f2685a/ruby-2.5/runtime.yaml#L43-L47)

> Let's build Python-3.7 function as an example:

Reviewer: ```suggestion
Let's build a Python 3.7 function as an example:
```

1. Create directory and save your function code there:

```
  mkdir python
  cd python
  cat > handler.py <<EOF
import json
import datetime
def endpoint(event, context):
  current_time = datetime.datetime.now().time()
  body = {
      "message": "Hello, the current time is " + str(current_time)
  }
  response = {
      "statusCode": 200,
      "body": json.dumps(body)
  }
  return response
EOF
```

> 2. Extract runtime [Dockerfile](https://github.com/triggermesh/knative-lambda-runtime/blob/9a74ce1ac03d56d233cfc7a46d84f2c5e5f2685a/python-3.7/runtime.yaml#L43-L50), update `_HANDLER` variable and store it in the same directory:

Reviewer: ```suggestion
2. Extract the runtime's [Dockerfile](https://github.com/triggermesh/knative-lambda-runtime/blob/9a74ce1ac03d56d233cfc7a46d84f2c5e5f2685a/python-3.7/runtime.yaml#L43-L50), store it in the same directory, and update the `_HANDLER` variable:
```

> 2. Extract runtime [Dockerfile](https://github.com/triggermesh/knative-lambda-runtime/blob/9a74ce1ac03d56d233cfc7a46d84f2c5e5f2685a/python-3.7/runtime.yaml#L43-L50), update `_HANDLER` variable and store it in the same directory:

Reviewer 2: same with that link

```
  cat > Dockerfile <<EOF
  FROM gcr.io/triggermesh/knative-lambda-python37
  ENV _HANDLER handler.endpoint
  COPY . .
  RUN if [ -f requirements.txt ]; then pip3.7 install -r requirements.txt ;fi
  ENTRYPOINT ["/opt/aws-custom-runtime"]
EOF
```

> `_HANDLER` variable in most cases consist of the filename without the file format part and function name.

Reviewer: ```suggestion
The `_HANDLER` variable in most cases consists of the filename without the file extension, and the function name.
```

> `_HANDLER` variable in most cases consist of the filename without the file format part and function name.

Reviewer 2: most cases ? what are the other cases ?

Author: It depends on the language and Lambda's implementation of the runtime. Here are some examples of uncommon cases:

- https://docs.aws.amazon.com/lambda/latest/dg/golang-handler.html
Golang: there is always must be the `main` package with the `main()` function, so in fact, Go runtime doesn't need a custom handler at all.
  
- https://docs.aws.amazon.com/lambda/latest/dg/java-handler.html
Java: Afaik, Java compiler puts everything in its paths and which only Java dev can figure out, i.e. handler may look like `com.serverless.Handler` and it will run code in `src/main/java/com/serverless/Handler.java` file. 

I really think that if someone wants to run his AWS Lambda functions in KLR he should be familiar with the "handler" thing. Otherwise, we should copy a big part of AWS documentation.

3. Build, run, test:

```
  docker build -t python-klr-image .
  docker run -d --rm --name python-klr-container python-klr-image
  # following command will work if you use Docker bridge network and you have jq tool 
  # otherwise, you should get the container address manually  
  curl $(docker inspect python-klr-container | jq .[].NetworkSettings.Networks.bridge.IPAddress -r):8080
```
  

>   The response will contain JSON with the current time.

Reviewer: ```suggestion
  The response will contain a JSON document with the current time.
```

4. Cleanup:

```

>   docker stop python-klr-container

Reviewer 2: and docker kill

Author: I thought that `stop` is `SIGTERM` -> 10s wait -> `SIGKILL`, while `kill` is just `SIGKILL`, but I may be wrong :thinking:

Reviewer: Agree, please never use kill unless absolutely necessary.

I think Seb meant `rm`.

Reviewer 2: yes my bad. I meant `rm`

Author: I knew that you'll ask about the `rm` and that's why I put it in the `run` command: 
```
docker run -d --rm --name python-klr-container python-klr-image
```
https://github.com/triggermesh/knative-lambda-runtime/pull/50/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R315

> --rm                             Automatically remove the container when it exits
